File: backend/src/controllers/authController.js
import jwt from 'jsonwebtoken';
import User from '../models/User.js';
import Match from '../models/Match.js';
import Message from '../models/Message.js';
import Expense from '../models/Expense.js';
import generateAnon from '../utils/generateAnon.js';

const genToken = (id) =>
  jwt.sign({ id }, process.env.JWT_SECRET, { expiresIn: '7d' });

// Register User
export const registerUser = async (req, res) => {
  const { name, email, phone, password, gender, isVerified, maskedAadhaar } = req.body;
  if (!name || !email || !password)
    return res.status(400).json({ msg: 'All fields required' });

  try {
    // Check if email or phone already exists
    const emailExists = await User.findOne({ email });
    if (emailExists) return res.status(400).json({ msg: 'Account already exists with this email address' });

    if (phone) {
      const phoneExists = await User.findOne({ phone });
      if (phoneExists) return res.status(400).json({ msg: 'Account already exists with this phone number' });
    }

    const avatarSeed = Math.random().toString(36).substring(7);

    const user = await User.create({
      name,
      email,
      phone,
      password,
      anonymousId: generateAnon(),
      gender,
      isVerified: !!isVerified,
      maskedAadhaar,
      avatarSeed
    });

    res.status(201).json({
      _id: user._id,
      name: user.name,
      email: user.email,
      anonymousId: user.anonymousId,
      token: genToken(user._id)
    });
  } catch (err) {
    res.status(500).json({ msg: 'Server error', err: err.message });
  }
};

// Login User
export const loginUser = async (req, res) => {
  const { email, password } = req.body;
  try {
    const user = await User.findOne({ email });
    if (!user || !(await user.matchPassword(password)))
      return res.status(401).json({ msg: 'Invalid credentials' });

    res.json({
      _id: user._id,
      name: user.name,
      email: user.email,
      anonymousId: user.anonymousId,
      token: genToken(user._id)
    });
  } catch (err) {
    res.status(500).json({ msg: 'Server error', err: err.message });
  }
};

// Google Login / Register
export const googleLogin = async (req, res) => {
  const { email, name, googleId } = req.body;
  if (!email || !name)
    return res.status(400).json({ msg: 'Email and name are required' });

  try {
    let user = await User.findOne({ email });
    if (!user) {
      return res.status(404).json({ msg: 'Account does not exist. Please sign up first!' });
    }

    res.json({
      _id: user._id,
      name: user.name,
      email: user.email,
      anonymousId: user.anonymousId,
      token: genToken(user._id)
    });
  } catch (err) {
    res.status(500).json({ msg: 'Server error', err: err.message });
  }
};

// Update Profile (e.g. Save Voice Assessment Vibe Description)
export const updateProfile = async (req, res) => {
  const { vibeText } = req.body;
  try {
    const user = await User.findById(req.user._id);
    if (!user) return res.status(404).json({ msg: 'User not found' });

    if (vibeText !== undefined) user.vibeText = vibeText;
    await user.save();

    res.json({
      _id: user._id,
      name: user.name,
      email: user.email,
      anonymousId: user.anonymousId,
      vibeText: user.vibeText
    });
  } catch (err) {
    res.status(500).json({ msg: 'Server error', err: err.message });
  }
};

// Get verified roommate candidates (excluding self, showing only anonymized profiles)
export const getCandidates = async (req, res) => {
  try {
    // Find all verified users who aren't the current user
    const users = await User.find({
      _id: { $ne: req.user._id },
      isVerified: true
    }).select('anonymousId vibeText avatarSeed');

    // Return the anonymized user records
    res.json(users);
  } catch (err) {
    res.status(500).json({ msg: 'Server error', err: err.message });
  }
};

// Like/Swipe a candidate
export const likeCandidate = async (req, res) => {
  const { candidateId, like } = req.body;
  if (!candidateId) return res.status(400).json({ msg: 'Candidate ID required' });

  try {
    const userA = req.user._id.toString();
    const userB = candidateId.toString();

    // Alphabetical ordering to keep index keys consistent
    const [firstUser, secondUser] = userA < userB ? [userA, userB] : [userB, userA];

    let match = await Match.findOne({ userA: firstUser, userB: secondUser });
    if (!match) {
      match = new Match({
        userA: firstUser,
        userB: secondUser,
        userALiked: userA === firstUser ? like : false,
        userBLiked: userA === secondUser ? like : false,
        status: 'pending'
      });
    } else {
      if (userA === firstUser) match.userALiked = like;
      if (userA === secondUser) match.userBLiked = like;

      if (match.userALiked && match.userBLiked) {
        match.status = 'matched';
      }
    }

    await match.save();
    res.json({
      matchStatus: match.status,
      isMatch: match.status === 'matched'
    });
  } catch (err) {
    res.status(500).json({ msg: 'Server error', err: err.message });
  }
};

// Get revealed matches
export const getMatches = async (req, res) => {
  try {
    const userId = req.user._id;
    // Find matches where current user is userA or userB and status is 'matched'
    const matches = await Match.find({
      $or: [{ userA: userId }, { userB: userId }],
      status: 'matched'
    }).populate('userA', 'name email anonymousId avatarSeed').populate('userB', 'name email anonymousId avatarSeed');

    const result = matches.map(m => {
      const otherUser = m.userA._id.toString() === userId.toString() ? m.userB : m.userA;
      return {
        _id: otherUser._id,
        name: otherUser.name,
        email: otherUser.email,
        anonymousId: otherUser.anonymousId,
        avatarSeed: otherUser.avatarSeed,
        status: m.status
      };
    });

    res.json(result);
  } catch (err) {
    res.status(500).json({ msg: 'Server error', err: err.message });
  }
};

// Chatterbox Chat - Send Message
export const sendMessage = async (req, res) => {
  const { receiverId, content } = req.body;
  if (!receiverId || !content)
    return res.status(400).json({ msg: 'Receiver ID and content are required' });

  try {
    const message = await Message.create({
      sender: req.user._id,
      receiver: receiverId,
      content
    });
    res.status(201).json(message);
  } catch (err) {
    res.status(500).json({ msg: 'Server error', err: err.message });
  }
};

// Chatterbox Chat - Get Message History
export const getMessages = async (req, res) => {
  const { recipientId } = req.params;
  try {
    const messages = await Message.find({
      $or: [
        { sender: req.user._id, receiver: recipientId },
        { sender: recipientId, receiver: req.user._id }
      ]
    }).sort({ timestamp: 1 });

    res.json(messages);
  } catch (err) {
    res.status(500).json({ msg: 'Server error', err: err.message });
  }
};

// Chatterbox Chat - Get Active Conversations Count
export const getActiveChatsCount = async (req, res) => {
  try {
    const messages = await Message.find({
      $or: [
        { sender: req.user._id },
        { receiver: req.user._id }
      ]
    });

    const activeUsers = new Set();
    messages.forEach(m => {
      if (m.sender.toString() !== req.user._id.toString()) {
        activeUsers.add(m.sender.toString());
      }
      if (m.receiver.toString() !== req.user._id.toString()) {
        activeUsers.add(m.receiver.toString());
      }
    });

    res.json({ count: activeUsers.size });
  } catch (err) {
    res.status(500).json({ msg: 'Server error', err: err.message });
  }
};

// Splits - Add shared expense
export const addExpense = async (req, res) => {
  const { amount, description, splitWith, category } = req.body;
  if (!amount || !description || !splitWith)
    return res.status(400).json({ msg: 'Amount, description and split roommate required' });

  try {
    const expense = await Expense.create({
      amount,
      description,
      paidBy: req.user._id,
      splitWith,
      category: category || 'General'
    });
    res.status(201).json(expense);
  } catch (err) {
    res.status(500).json({ msg: 'Server error', err: err.message });
  }
};

// Splits - Get shared expenses between current user and roommate
export const getExpenses = async (req, res) => {
  const { roommateId } = req.query;
  if (!roommateId) return res.status(400).json({ msg: 'Roommate ID required' });

  try {
    const userId = req.user._id;
    const expenses = await Expense.find({
      $or: [
        { paidBy: userId, splitWith: roommateId },
        { paidBy: roommateId, splitWith: userId }
      ]
    }).sort({ date: -1 });

    res.json(expenses);
  } catch (err) {
    res.status(500).json({ msg: 'Server error', err: err.message });
  }
};

File: backend/src/main/java/com/intune/backend/controller/AuthController.java
package com.intune.backend.controller;

import com.intune.backend.model.Match;
import com.intune.backend.model.User;
import com.intune.backend.repository.MatchRepository;
import com.intune.backend.repository.UserRepository;
import com.intune.backend.security.JwtTokenProvider;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.client.RestTemplate;

import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private MatchRepository matchRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtTokenProvider tokenProvider;

    @Value("${ai.similarity.url}")
    private String aiSimilarityUrl;

    private final RestTemplate restTemplate = new RestTemplate();

    // DTOs using Lombok for clean, readable code
    @Data
    public static class RegisterRequest {
        private String name;
        private String email;
        private String phone;
        private String password;
        private String gender;
        private boolean isVerified;
        private String maskedAadhaar;
        private String aadhaarNumber;
    }

    @Data
    public static class LoginRequest {
        private String email;
        private String password;
    }

    @Data
    public static class GoogleLoginRequest {
        private String email;
        private String name;
        private String googleId;
    }

    @Data
    public static class ProfileRequest {
        private String vibeText;
    }

    @Data
    public static class LikeRequest {
        private String candidateId;
        private boolean like;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class CandidateResponse {
        private String _id;
        private String anonymousId;
        private String vibeText;
        private String avatarSeed;
        private double match_score;
        private boolean isNewMatch;
    }

    // Helper to generate anonymous ID
    private String generateAnon() {
        String[] adjectives = {"Sky", "Moon", "Star", "Sun", "River", "Cloud"};
        String word = adjectives[(int) (Math.random() * adjectives.length)];
        int num = (int) (100 + Math.random() * 900); // 100-999
        return word + "_" + num;
    }

    private String hashAadhaar(String aadhaar) {
        if (aadhaar == null || aadhaar.trim().isEmpty()) return null;
        try {
            java.security.MessageDigest digest = java.security.MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(aadhaar.trim().getBytes(java.nio.charset.StandardCharsets.UTF_8));
            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) hexString.append('0');
                hexString.append(hex);
            }
            return hexString.toString();
        } catch (Exception e) {
            throw new RuntimeException("SHA-256 initialization error", e);
        }
    }

    @PostMapping("/register")
    public ResponseEntity<?> registerUser(@RequestBody RegisterRequest request) {
        if (request.getEmail() == null || request.getPassword() == null || request.getName() == null) {
            return ResponseEntity.badRequest().body(Map.of("msg", "All fields required"));
        }

        if (userRepository.findByEmail(request.getEmail()).isPresent()) {
            return ResponseEntity.badRequest().body(Map.of("msg", "Account already exists with this email address"));
        }

        if (request.getPhone() != null && !request.getPhone().isEmpty()) {
            if (userRepository.findByPhone(request.getPhone()).isPresent()) {
                return ResponseEntity.badRequest().body(Map.of("msg", "Account already exists with this phone number"));
            }
        }

        String aadhHash = null;
        if (request.getAadhaarNumber() != null && !request.getAadhaarNumber().trim().isEmpty()) {
            aadhHash = hashAadhaar(request.getAadhaarNumber());
            if (userRepository.findByAadhaarHash(aadhHash).isPresent()) {
                return ResponseEntity.badRequest().body(Map.of("msg", "An account is already verified with this Aadhaar number"));
            }
        }

        String avatarSeed = UUID.randomUUID().toString().substring(0, 7);

        User user = User.builder()
                .name(request.getName())
                .email(request.getEmail())
                .phone(request.getPhone())
                .password(passwordEncoder.encode(request.getPassword()))
                .anonymousId(generateAnon())
                .gender(request.getGender())
                .isVerified(request.isVerified())
                .maskedAadhaar(request.getMaskedAadhaar())
                .aadhaarHash(aadhHash)
                .avatarSeed(avatarSeed)
                .createdAt(new Date())
                .updatedAt(new Date())
                .build();

        userRepository.save(user);

        String token = tokenProvider.generateToken(user.getId());

        return ResponseEntity.status(HttpStatus.CREATED).body(Map.of(
                "_id", user.getId(),
                "name", user.getName(),
                "email", user.getEmail(),
                "anonymousId", user.getAnonymousId(),
                "token", token
        ));
    }

    @PostMapping("/login")
    public ResponseEntity<?> loginUser(@RequestBody LoginRequest request) {
        Optional<User> userOpt = userRepository.findByEmail(request.getEmail());
        if (userOpt.isEmpty() || !passwordEncoder.matches(request.getPassword(), userOpt.get().getPassword())) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(Map.of("msg", "Invalid credentials"));
        }

        User user = userOpt.get();
        String token = tokenProvider.generateToken(user.getId());

        return ResponseEntity.ok(Map.of(
                "_id", user.getId(),
                "name", user.getName(),
                "email", user.getEmail(),
                "anonymousId", user.getAnonymousId(),
                "token", token
        ));
    }

    @PostMapping("/google-login")
    public ResponseEntity<?> googleLogin(@RequestBody GoogleLoginRequest request) {
        Optional<User> userOpt = userRepository.findByEmail(request.getEmail());
        if (userOpt.isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("msg", "Account does not exist. Please sign up first."));
        }

        User user = userOpt.get();
        String token = tokenProvider.generateToken(user.getId());

        return ResponseEntity.ok(Map.of(
                "_id", user.getId(),
                "name", user.getName(),
                "email", user.getEmail(),
                "anonymousId", user.getAnonymousId(),
                "token", token
        ));
    }

    @GetMapping("/me")
    public ResponseEntity<?> getMe() {
        User user = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        return ResponseEntity.ok(user);
    }

    @PutMapping("/profile")
    public ResponseEntity<?> updateProfile(@RequestBody ProfileRequest request) {
        User currentUser = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        
        Optional<User> userOpt = userRepository.findById(currentUser.getId());
        if (userOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("msg", "User not found"));
        }

        User user = userOpt.get();
        if (request.getVibeText() != null) {
            user.setVibeText(request.getVibeText());
        }
        user.setUpdatedAt(new Date());
        userRepository.save(user);

        return ResponseEntity.ok(user);
    }

    @GetMapping("/candidates")
    public ResponseEntity<?> getCandidates() {
        User currentUser = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();

        // Query all verified candidate users excluding self
        List<User> candidates = userRepository.findByIdNot(currentUser.getId()).stream()
                .filter(User::isVerified)
                .collect(Collectors.toList());

        if (candidates.isEmpty()) {
            return ResponseEntity.ok(Collections.emptyList());
        }

        // Gather all candidate vibe texts
        List<String> candidateVibes = candidates.stream()
                .map(c -> c.getVibeText() != null && !c.getVibeText().isEmpty() ? c.getVibeText() : "No vibe details provided.")
                .collect(Collectors.toList());

        List<Double> scores = new ArrayList<>();
        boolean success = false;

        // Hit the SBERT FastAPI similarity service
        if (currentUser.getVibeText() != null && !currentUser.getVibeText().trim().isEmpty()) {
            try {
                Map<String, Object> payload = Map.of(
                        "anchor", currentUser.getVibeText(),
                        "candidates", candidateVibes
                );
                Map<String, Object> response = restTemplate.postForObject(aiSimilarityUrl, payload, Map.class);
                if (response != null && response.containsKey("scores")) {
                    List<?> scoresList = (List<?>) response.get("scores");
                    for (Object scoreObj : scoresList) {
                        scores.add(((Number) scoreObj).doubleValue());
                    }
                    success = true;
                }
            } catch (Exception e) {
                // Fallback to local similarity mapping if SBERT is offline
            }
        }

        // Fallback calculations if SBERT fails/is offline
        if (!success) {
            for (User cand : candidates) {
                scores.add(calculateFallbackSimilarity(currentUser.getVibeText(), cand.getVibeText()));
            }
        }

        // Build responses
        List<CandidateResponse> responseList = new ArrayList<>();
        long oneDayAgoMs = System.currentTimeMillis() - (24 * 60 * 60 * 1000L); // 24 hours threshold

        for (int i = 0; i < candidates.size(); i++) {
            User cand = candidates.get(i);
            double score = scores.get(i);
            
            // Check if user is newly registered (within last 24h) and highly compatible (score >= 80%)
            boolean isNewMatch = false;
            if (cand.getCreatedAt() != null && cand.getCreatedAt().getTime() > oneDayAgoMs) {
                if (score >= 80.0) {
                    isNewMatch = true;
                }
            }

            responseList.add(CandidateResponse.builder()
                    ._id(cand.getId())
                    .anonymousId(cand.getAnonymousId())
                    .vibeText(cand.getVibeText())
                    .avatarSeed(cand.getAvatarSeed())
                    .match_score(score)
                    .isNewMatch(isNewMatch)
                    .build());
        }

        // Sort candidates by match score in descending order
        responseList.sort((a, b) -> Double.compare(b.getMatch_score(), a.getMatch_score()));

        return ResponseEntity.ok(responseList);
    }

    @PostMapping("/like")
    public ResponseEntity<?> likeCandidate(@RequestBody LikeRequest request) {
        if (request.getCandidateId() == null) {
            return ResponseEntity.badRequest().body(Map.of("msg", "Candidate ID required"));
        }

        User currentUser = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        String userA = currentUser.getId();
        String userB = request.getCandidateId();

        // Maintain consistent alphabetical sorting for unique matches table compound index
        String firstUser = userA.compareTo(userB) < 0 ? userA : userB;
        String secondUser = userA.compareTo(userB) < 0 ? userB : userA;

        Optional<Match> matchOpt = matchRepository.findByUserAAndUserB(firstUser, secondUser);
        Match match;

        if (matchOpt.isEmpty()) {
            match = Match.builder()
                    .userA(firstUser)
                    .userB(secondUser)
                    .userALiked(userA.equals(firstUser) ? request.isLike() : false)
                    .userBLiked(userA.equals(secondUser) ? request.isLike() : false)
                    .status("pending")
                    .createdAt(new Date())
                    .updatedAt(new Date())
                    .build();
        } else {
            match = matchOpt.get();
            if (userA.equals(firstUser)) match.setUserALiked(request.isLike());
            if (userA.equals(secondUser)) match.setUserBLiked(request.isLike());

            if (match.isUserALiked() && match.isUserBLiked()) {
                match.setStatus("matched");
            }
            match.setUpdatedAt(new Date());
        }

        matchRepository.save(match);

        return ResponseEntity.ok(Map.of(
                "matchStatus", match.getStatus(),
                "isMatch", match.getStatus().equals("matched")
        ));
    }

    @GetMapping("/matches")
    public ResponseEntity<?> getMatches() {
        User currentUser = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        String userId = currentUser.getId();

        List<Match> matches = matchRepository.findConfirmedMatchesForUser(userId);

        List<Map<String, Object>> resultList = new ArrayList<>();
        for (Match match : matches) {
            String otherUserId = match.getUserA().equals(userId) ? match.getUserB() : match.getUserA();
            Optional<User> otherUserOpt = userRepository.findById(otherUserId);
            if (otherUserOpt.isPresent()) {
                User otherUser = otherUserOpt.get();
                resultList.add(Map.of(
                        "_id", otherUser.getId(),
                        "name", otherUser.getName(),
                        "email", otherUser.getEmail(),
                        "anonymousId", otherUser.getAnonymousId(),
                        "avatarSeed", otherUser.getAvatarSeed() != null ? otherUser.getAvatarSeed() : "",
                        "status", match.getStatus()
                ));
            }
        }

        return ResponseEntity.ok(resultList);
    }

    // Jaccard similarity fallback calculation mapped to [55, 98] range
    private double calculateFallbackSimilarity(String text1, String text2) {
        if (text1 == null || text2 == null || text1.trim().isEmpty() || text2.trim().isEmpty()) {
            return 50.0;
        }

        Set<String> words1 = new HashSet<>(Arrays.asList(text1.toLowerCase().split("\\W+")));
        Set<String> words2 = new HashSet<>(Arrays.asList(text2.toLowerCase().split("\\W+")));

        Set<String> intersection = new HashSet<>(words1);
        intersection.retainAll(words2);

        Set<String> union = new HashSet<>(words1);
        union.addAll(words2);

        if (union.isEmpty()) return 50.0;

        double jaccard = (double) intersection.size() / union.size();
        return 55.0 + (jaccard * 43.0);
    }
}
